File: NET/JNetCall.Sharp/Client/ServiceClient.cs
using System;
using Castle.DynamicProxy;

namespace JNetCall.Sharp.Client
{
    public static class ServiceClient
    {
        private static readonly Lazy<ProxyGenerator> Generator;

        static ServiceClient()
        {
            Generator = new Lazy<ProxyGenerator>(() => new ProxyGenerator());
        }

        public static T CreateNative<T>(string exe) where T : class
        {
            var interceptor = new JvmInterceptor(exe);
            return Create<T>(interceptor);
        }

        public static T CreateMain<T>(string exe) where T : class
        {
            var interceptor = new JavaInterceptor(exe);
            return Create<T>(interceptor);
        }

        private static T Create<T>(IInterceptor interceptor) where T : class
        {
            var gen = Generator.Value;
            var proxy = gen.CreateInterfaceProxyWithoutTarget<T>(interceptor);
            return proxy;
        }
    }
}

File: JVM/jnetcall-java/src/main/java/jnetcall/java/client/ServiceClient.java
package jnetcall.java.client;

import java.lang.reflect.InvocationHandler;
import java.lang.reflect.Proxy;

public final class ServiceClient {

    public static <T> T createNative(Class<T> clazz, String exe) {
        var handler = new ClrInterceptor(exe);
        return create(clazz, handler);
    }

    public static <T> T createMain(Class<T> clazz, String exe) {
        var handler = new NetInterceptor(exe);
        return create(clazz, handler);
    }

    private static <T> T create(Class<T> clazz, InvocationHandler handler) {
        var loader = clazz.getClassLoader();
        T proxy = (T) Proxy.newProxyInstance(loader, new Class[]{clazz}, handler);
        return proxy;
    }
}
